### PR TITLE
Simplify setup.py and move towards cookie-cutter configuration

### PR DESCRIPTION
**Pull request by agoose77**
This PR bumps the dependency versions of the JS package to follow the released versions rather than RCs.
It also further simplifies setup.py to follow the structure of the cookie-cutter used for prebuilt extensions.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -11,45 +8,29 @@
...
     data_files_spec=[
-        ('share/jupyter/nbextensions/jupyter-threejs',
-         name + '/static',
-         '*.js'),
-        ('share/jupyter/nbextensions/jupyter-threejs',
-         name + '/static',
-         '*.js.map'),
-        ('share/jupyter/lab/extensions',
-         'js/lab-dist',
-         'jupyter-threejs-*.tgz'),
```
Comment (maartenbreddels) [inline setup.py:36]: i guess we still want lab2 support?

File: js/package.json [hunk 2]
```diff
@@ -54,7 +54,7 @@
   },
   "jupyterlab": {
     "extension": "src/jupyterlab-plugin",
-    "outputDir": "../share/jupyter/labextensions/jupyter-threejs",
+    "outputDir": "../pythreejs/labextension",
```
Comment (maartenbreddels) [inline js/package.json:57]: This doesn't match the `"share/jupyter/labextensions/jupyter-threejs"` in setup.py right?

**Review by maartenbreddels — COMMENTED** [review]
Thnx for picking this up!

File: js/package.json (hunk 2 above)
Comment (agoose77) [reply js/package.json:57]: Ah, looks like some of my changes have mixed up... I'll resolve this

Comment (agoose77) [reply js/package.json:57]: I think this is correct: the outputdir maps to `<ROOT>/pythreejs/labextension`, and `setup.py` looks for `<ROOT>/pythreejs/labextension`

File: setup.py (hunk 1 above)
Comment (agoose77) [reply setup.py:36]: Yes, I suppose we do (please excuse the slash-and-burn approach :P)

**Comment by agoose77** [comment]
@maartenbreddels I've restored support for JupyterLab2 extensions, and replied to your query about the extension; both the setup.py and package.json refer to the same (<ROOT>/pythreejs/labextension) directory

File: setup.py [hunk 3]
```diff
@@ -11,45 +8,31 @@
...
-        ('share/jupyter/labextensions/jupyter-threejs/',
-         'share/jupyter/labextensions/jupyter-threejs/',
-         '*.*'),
-        ('share/jupyter/labextensions/jupyter-threejs/static',
-         'share/jupyter/labextensions/jupyter-threejs/static/',
-         '*.*'),
-        ('etc/jupyter/nbconfig',
-         'jupyter-config',
-         '**/*.json'),
```
Comment (maartenbreddels) [inline setup.py:45]: why is this removed, i guess we still want that right?

File: js/package.json (hunk 2 above)
Comment (maartenbreddels) [inline js/package.json:57]: I don't think this makes things simpler to understand, having the same directory keeps things simple.

Comment (agoose77) [reply js/package.json:57]: OK, will revert this :)

File: setup.py (hunk 3 above)
Comment (bollwyvl) [reply setup.py:45]: yeah, should keep.

File: js/package.json [hunk 4]
```diff
@@ -34,14 +34,14 @@
...
-    "@jupyter-widgets/base": "^1.2.5 || ^2.0.0 || ^3.0.0 || ^4.0.0-alpha.2",
+    "@jupyter-widgets/base": "^1.2.5 || ^2.0.0 || ^3.0.0 || ^4.0.0",
     "bluebird": "^3.5.5",
     "jupyter-dataserializers": "^2.2.0",
     "three": "^0.97.0",
     "underscore": "^1.8.3"
   },
   "devDependencies": {
-    "@jupyterlab/builder": "^3.0.0-rc.13",
+    "@jupyterlab/builder": "^3.0.0",
```
Comment (bollwyvl) [inline js/package.json:44]: needs bump to at least `3.0.1`

File: setup.py (hunk 3 above)
Comment (bollwyvl) [reply setup.py:45]: I put this under test on the [github actions PR](https://github.com/jupyter-widgets/pythreejs/pull/349/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR171)

File: setup.py [hunk 5]
```diff
@@ -91,7 +74,7 @@
             'sphinx-rtd-theme',
         ]
     },
-    'packages': [name],  # Manually specify here, update after autogen
+    'packages': setuptools.find_packages(),
```
Comment (bollwyvl) [inline setup.py]: this is always a bit fraught... it would be good to check exactly what goes in, and potentially add some `exclude`. Moving to a `src/` layout helps this _a lot_, but is probably out of scope for this.

File: setup.py (hunk 3 above)
Comment (agoose77) [reply setup.py:45]: Yes, this was an accidental remove :/ Will fix.

**Comment by bollwyvl** [comment]
i still don't know what it does for us, but `install.json` seems to be showing up in other places. more metadata good!

File: js/package.json (hunk 2 above)
Comment (bollwyvl) [reply js/package.json:57]: the only concern might be path length, if it's getting deep, for windows folks. 256 chars comes up so fast... and frigging CI machines already have it set, so it's up to users to discover...

**Comment by agoose77** [comment]
I'll address the comments tomorrow GMT :)

File: js/package.json (hunk 4 above)
Comment (agoose77) [reply js/package.json:44]: @bollwyvl what's the issue with the 3.0.0 release?

File: js/package.json (hunk 2 above)
Comment (agoose77) [reply js/package.json:57]: I'm not a windows user these days, but I can appreciate the concern; @maartenbreddels do you feel like comfortable leaving it in-tree?

File: setup.py (hunk 3 above)
Comment (agoose77) [reply setup.py:45]: Resolved by https://github.com/maartenbreddels/pythreejs/pull/2/commits/8b0a734e78f0f1568b23904d8e5de0d37d73d71a

File: setup.py (hunk 5 above)
Comment (agoose77) [reply setup.py]: Yeah, let's push any major layout changes that a different PR :)

File: js/package.json (hunk 2 above)
Comment (maartenbreddels) [reply js/package.json:57]: ok, lets leave it as is.

File: js/package.json (hunk 4 above)
Comment (bollwyvl) [reply js/package.json:44]: The last time i was working down at this level, it complained to me when building... but i can't find where it applies the check in core...

File: setup.py (hunk 5 above)
Comment (vidartf) [reply setup.py]: Was this motivated by something? I think the comment is still important, as *the package list made here won't include the autogenerated python subpackages* no matter which method we choose. I see the code of the pythreejs PR simply removed this important step: https://github.com/jupyter-widgets/pythreejs/pull/342#discussion_r575259828

Comment (agoose77) [reply setup.py]: I seem to have dropped the ball on this; I tested things locally but something must have been cached. I'll revert the change.

**Comment by maartenbreddels** [comment]
why is https://github.com/maartenbreddels/pythreejs/pull/2/commits/5eee7d1651d4178b5e5e271f646c5bc7fbbc65f1 needed? I'm afraid if we don't keep this PR minimal, we will never ship this.

**Comment by agoose77** [comment]
@maartenbreddels it's a fair concern!

I made an mistake when removing the bundled `setupbase.py` in favour of the `jupyter_packaging` one. I'm not really sure what happened because I tested everything locally, but I must have missed something. The last few commits https://github.com/maartenbreddels/pythreejs/pull/2/commits/8b0a734e78f0f1568b23904d8e5de0d37d73d71a ... https://github.com/maartenbreddels/pythreejs/pull/2/commits/adc106ec87666cabda8799a2790d8684ef8ab379 are all to fix these oversights. I kept them in separate commits to keep the history clean.

In particular, https://github.com/maartenbreddels/pythreejs/commit/5eee7d1651d4178b5e5e271f646c5bc7fbbc65f1 just restores the `pythreejs` specific change to `setupbase.py` by monkeypatching it. Unfortunately there isn't a cleaner way to do this as far as I can see, and I think it's important to avoid having a separate implementation of `jupyter_packaging`

The [files view](https://github.com/maartenbreddels/pythreejs/pull/2/files) gives a better overview of how simple this PR is! :smile:

**Comment by maartenbreddels** [comment]
Can you revert that removal?

(from mobile phone)

**Comment by agoose77** [comment]
I can, would you rather bundle an entire `setupbase.py` vs depending upon `jupyter_packaging`?

**Comment by maartenbreddels** [comment]
I'd rather see that go in a different PR if it gives trouble, so we can
focus on jlab3 in this pr.

(from mobile phone)

**Comment by agoose77** [comment]
OK, thanks for all the input everyone. I'm going to close this, because the revert would need to happen in your branch (https://github.com/jupyter-widgets/pythreejs/pull/342/commits/df6afcfa0e82f3cd9d996526f672b6e48e8f6298)... and the history is getting messy! I will open a new PR on jupyter-widgets/pythreejs  that builds on top of 49ffb97
